src/lib/state_or_error/state_or_error.ml

```diff
@@ -0,0 +1,103 @@
+open Core_kernel
+
+module Make : State_or_error_intf.S =
+functor
+  (State : State_or_error_intf.State_intf)
+  ->
+  struct
+    module T = struct
+      type state = State.t
+
+      type 'a t = state -> ('a * state) Or_error.t
+
+      let return : 'a -> 'a t = fun a s -> Ok (a, s)
+
+      let bind m ~f = function
+        | s ->
+            let open Or_error.Let_syntax in
+            let%bind a, s' = m s in
+            f a s'
+
+      let map = `Define_using_bind
+    end
+
+    include T
+    include Monad.Make (T)
+
+    let get = function s -> Ok (s, s)
+
+    let put s = function _ -> Ok ((), s)
+
+    let run_state t ~state = t state
+
+    let error_if b ~message ~value =
+      if b then fun _ -> Or_error.error_string message else return value
+  end
+
+module Make2 : State_or_error_intf.S2 =
+functor
+  (State : State_or_error_intf.State_intf1)
+  ->
+  struct
+    module T = struct
+      type 'a state = 'a State.t
+
+      type ('b, 'a) t = 'a state -> ('b * 'a state) Or_error.t
+
+      let return : 'b -> ('b, 'a) t = fun a s -> Ok (a, s)
+
+      let bind m ~f = function
+        | s ->
+            let open Or_error.Let_syntax in
+            let%bind a, s' = m s in
+            f a s'
+
+      let map = `Define_using_bind
+    end
+
+    include T
+    include Monad.Make2 (T)
+
+    let get = function s -> Ok (s, s)
+
+    let put s = function _ -> Ok ((), s)
+
+    let run_state t ~state = t state
+
+    let error_if b ~message ~value =
+      if b then fun _ -> Or_error.error_string message else return value
+  end
+
+module Make3 : State_or_error_intf.S3 =
```

Comment:
You don't have to write this functor 3 times, there are converter functors. You can write `Make3` and use `S2_to_S3` to write `Make2` and use `S_to_S2` to write `Make`.

https://ocaml.janestreet.com/ocaml-core/latest/doc/base/Base__/Monad_intf/index.html#module-S3_to_S2